pkg/apis/core/validation/validation_patch_test.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/kubernetes/pkg/apis/core"
+)
+
+func TestOpenShiftValidateSecretUpdate(t *testing.T) {
+	newSecretFn := func(ns, name string, secretType core.SecretType) *core.Secret {
+		return &core.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            name,
+				Namespace:       ns,
+				ResourceVersion: "1",
+			},
+			Type: secretType,
+			Data: map[string][]byte{
+				"tls.key": []byte("foo"),
+				"tls.crt": []byte("bar"),
+			},
+		}
+	}
+	errFn := func(secretType core.SecretType) field.ErrorList {
+		return field.ErrorList{
+			field.Invalid(field.NewPath("type"), secretType, "field is immutable"),
+		}
+	}
+
+	t.Run("verify whitelist", func(t *testing.T) {
+		for key := range whitelist {
+			split := strings.Split(key, "/")
+			if len(split) != 2 {
+				t.Errorf("invalid key in whitelist - %s", key)
+				continue
+			}
+			ns, name := split[0], split[1]
+
+			// exercise a valid type mutation: "SecretTypeTLS" -> "kubernetes.io/tls"
+			oldSecret, newSecret := newSecretFn(ns, name, "SecretTypeTLS"), newSecretFn(ns, name, core.SecretTypeTLS)
+			if errs := ValidateSecretUpdate(newSecret, oldSecret); len(errs) > 0 {
+				t.Errorf("unexpected error: %v", errs)
+			}
+
+			// no type change, no validation failure expected
+			oldSecret, newSecret = newSecretFn(ns, name, core.SecretTypeTLS), newSecretFn(ns, name, core.SecretTypeTLS)
+			if errs := ValidateSecretUpdate(newSecret, oldSecret); len(errs) > 0 {
+				t.Errorf("unexpected error: %v", errs)
+			}
+
+			// exercise an invalid type mutation, we expect validation failure
+			errExpected := errFn(core.SecretTypeOpaque)
+			oldSecret, newSecret = newSecretFn(ns, name, "SecretTypeTLS"), newSecretFn(ns, name, core.SecretTypeOpaque)
+			if errGot := ValidateSecretUpdate(newSecret, oldSecret); !cmp.Equal(errExpected, errGot) {
+				t.Errorf("expected error: %v, diff: %s", errExpected, cmp.Diff(errExpected, errGot))
+			}
```

Comment:
I don't doubt that it works but I would like to see a test that shows the kubernetes.io/tls key validation is still enforced on these updates.